src/rulesets/Base/array.jl

```diff
@@ -43,32 +51,81 @@ function rrule(::typeof(Base.vect), X::Vararg{Any,N}) where {N}
     return Base.vect(X...), vect_pullback
 end
 
+"""
+    _make_real_zeros(xdots, xs)
+
+Forward rules for `vect` or `cat` may receive a mixture of data and `ZeroTangent`s.
+To avoid `vect(1, ZeroTangent(), 3)` or `hcat([1,2], ZeroTangent())`, this materialises
+each zero `xdot` to be `zero(x)`.
+"""
+_make_real_zeros(xdots, xs) = map(_real_zero, xdots, xs)
+_real_zero(xdot, x) = xdot
+_real_zero(xdot::AbstractZero, x) = zero(x)
```

Comment:
I wonder if this shouldn't be:
```suggestion
_real_zero(xdot::ZeroTangent, x) = zero(x)
_real_zero(xdot::DoesNotExist, x) = isapplicable(zero, x) ? zero(x) : xdot
```
So that we never end up calling `zero("abc")`

Reply:
So the example is something like `gradient(x -> ["abc", x][end], 1)`. That does actually work in Zygote; it fails in Diffractor seemingly before hitting `_real_zero`. 

What's DoesNotExist? I think `hasmethod` is pretty slow; maybe better to leave an error until we have a better plan?
```
julia> @btime Base.hasmethod(zero, Tuple{typeof($([1,2,3]))})
  min 293.075 ns, mean 306.452 ns (3 allocations, 144 bytes)
true
```

Reply:
oops `DoesNotExist` was renamed to `NoTangent`

We can push `hasmethod` to compile-time if we want with Tricks.jl.
If we really want.

but also maybe we want to do:
```suggestion
_real_zero(xdot::ZeroTangent, x) = zero(x)
_real_zero(xdot::NoTangent, x) = xdot
```

but yeah i think fine enough to leave it as an error til it becomes a problem.